beetsplug/lyrics.py

```diff
@@ -659,26 +671,133 @@ def commands(self):
+                ui.print_(u'RST files generated. to build, use one of:')
+                ui.print_(u'  sphinx-build -b html %s _build/html'
+                          % opts.writerst)
+                ui.print_(u'  sphinx-build -b epub %s _build/epub'
+                          % opts.writerst)
+                ui.print_((u'  sphinx-build -b latex %s _build/latex '
+                           u'&& make -C _build/latex all-pdf')
+                          % opts.writerst)
         cmd.func = func
         return [cmd]
 
+    def writerst(self, directory, item):
+        """Write the item to an RST file
+
+        this will keep state (in the `rst` variable) in order to avoid
+        writing continuously to the same files
+        """
+        if item is None or self.artist != item.artist:
+            if self.rst is not None:
+                slug = re.sub(r'\W+', '-', unidecode(self.artist).lower())
+                path = os.path.join(directory, 'artists', slug + u'.rst')
+                with open(path, 'wb') as output:
+                    output.write(self.rst.encode('utf-8'))
+                self.rst = None
+                if item is None:
+                    return
+            self.artist = item.artist
+            self.rst = u"%s\n%s\n\n.. contents::\n   :local:\n\n" \
+                       % (self.artist, u'=' * len(self.artist))
+        if self.album != item.album:
+            tmpalbum = self.album = item.album
+            if self.album == '':
+                tmpalbum = u'Unknown album'
+            self.rst += u"%s\n%s\n\n" % (tmpalbum,
+                                         u'-' * len(tmpalbum))
+        title_str = u":index:`%s`" % item.title
+        block = u'| ' + item.lyrics.replace(u'\n', u'\n| ')
+        self.rst += u"%s\n%s\n\n%s\n" % (title_str,
+                                         u'~' * len(title_str),
+                                         block)
+
+    def writerst_indexes(self, directory):
+        """Write conf.py and index.rst files necessary for Sphinx
+
+        We write minimal configurations that are necessary for Sphinx
+        to operate. We do not overwrite existing files so that
+        customizations are respected."""
+        try:
+            os.makedirs(os.path.join(directory, 'artists'))
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                pass
+            else:
+                raise
+        indexfile = os.path.join(directory, 'index.rst')
+        if not os.path.exists(indexfile):
+            with open(indexfile, 'wb') as output:
+                output.write(u'''Lyrics
+======
+
+* :ref:`Song index <genindex>`
+* :ref:`search`
+
+Artist index:
+
+.. toctree::
+   :maxdepth: 1
+   :glob:
+
+   artists/*
+''')
+        conffile = os.path.join(directory, 'conf.py')
+        if not os.path.exists(conffile):
+            with open(conffile, 'wb') as output:
+                output.write(u'''# -*- coding: utf-8 -*-
```

Comment:
This (and the similar stanza above) actually breaks on Python 3 (and would break on Python 2 if this string were non-ASCII). When a file is opened in binary mode (with `b` in the flags to `open`), then you can only write raw bytes to the file, not Unicode strings like `u'''...'''`. Here's the error, FWIW:

```
Traceback (most recent call last):
  File "beet", line 23, in <module>
    beets.ui.main()
  File "/Users/asampson/Documents/code/beets/beets/ui/__init__.py", line 1256, in main
    _raw_main(args)
  File "/Users/asampson/Documents/code/beets/beets/ui/__init__.py", line 1243, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/Users/asampson/Documents/code/beets/beetsplug/lyrics.py", line 694, in func
    self.writerst_indexes(opts.writerst)
  File "/Users/asampson/Documents/code/beets/beetsplug/lyrics.py", line 779, in writerst_indexes
    ''')
TypeError: a bytes-like object is required, not 'str'
```

The right resolution is probably to do an `.encode('utf8')` on the string before writing it.